### DISCUSSION TITLE
[General] How to deploy contracts under reproducible addresses?

### DISCUSSION THREAD

**Discussion by CodeSandwich** [zkSync-Community-Hub/zksync-developers#824] [General] [answered]
### Team or Project

Drips

### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

I need a stable address derivation scheme, so on EVM I would deploy the logic contract and then CREATE2 clones of them, which gives me a known derivation path. On zkSync I can't use clones, but I can cheaply re-deploy an existing contract to a new address, which is comparably good. I do it by obtaining the codehash of the "master copy" of the contract and then by passing it to ContractDeployer.create2. It works fine, but in the Foundry tests context it only works in the top-level tests contracts. Invoking this routine from any other contract causes the This method require system call flag error. I couldn't find anything about the system call flag in the docs or in the help channels. What is this and how to work this issue around? Is my Foundry tests contract a system contract?

**Comment by alhenaai** [comment]
In ZKsync, the deployment of contracts, especially using the `create2` method for reproducible addresses, involves specific mechanisms that differ from Ethereum. The error you're encountering, "This method require system call flag," suggests that there's an issue with how the contract deployment is being invoked in your Foundry tests.

Here are a few points to consider and steps you might take to resolve this issue:

1. **System Contract Calls**: In ZKsync, certain operations, like those involving `create2`, are restricted to system contracts or require specific flags due to the security model and the way system contracts interact with the ZKsync protocol. This might be why you're seeing the error when calling from non-top-level contracts.

2. **Deployment from Non-System Contracts**: If your Foundry tests contract is not recognized as a system contract, it may not have the necessary permissions to invoke `create2` directly. This is likely the cause of the "system call flag" error.

3. **Workaround**: To work around this, ensure that any contract that needs to deploy another contract using `create2` is either:
   - A system contract itself.
   - Has the appropriate permissions or flags set to allow it to make system-level calls.

4. **Documentation and Support**: Since this seems to be a nuanced issue specific to ZKsync's implementation and your development environment, I recommend discussing this further with the ZKsync developer community. You can raise this issue on the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions) for more in-depth support and potential solutions from other developers who might have faced similar challenges.

This approach should help you address the specific error and continue with your contract deployment strategy on ZKsync.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by CodeSandwich** [comment]
> Has the appropriate permissions or flags set to allow it to make system-level calls.

How do I achieve that?

**Reply by dutterbutter** [reply]
This should be properly documented so will update docs to include but you need to add:

```
[profile.default.zksync] 
# Enable EraVM extensions (ex system-mode)
enable_eravm_extensions = true
```

or via the CLI.

**Reply by CodeSandwich** [reply]
It doesn't seem to be doing anything in my case. This is my current `founry.toml`:
```toml
[profile.default]
solc_version = '0.8.20'
evm_version = 'shanghai'
optimizer_runs = 7_700
via_ir = true
verbosity = 1
fuzz_runs = 5
fs_permissions = [{ access = "read-write", path = "./"}]
[profile.default.zksync]
enable_eravm_extensions = true
[fmt]
line_length = 100
[fuzz]
runs = 4
```
I tried adding `enable_eravm_extensions = true` to plain `profile.default` too and passing `--zk-enable-eravm-extensions` to `forge test`, same result.

Here's the minimal reproduction of the problem:
```solidity
// SPDX-License-Identifier: GPL-3.0-only
pragma solidity ^0.8.20;

import {console, Test} from "forge-std/Test.sol";

IContractDeployer constant CONTRACT_DEPLOYER = IContractDeployer(address(0x8006));

interface IContractDeployer {
        function create2(
        bytes32 _salt,
        bytes32 _bytecodeHash,
        bytes calldata _input
    ) external payable returns (address newAddress);
}

contract DummyDeployerTest is Test {
    function testDummyDeployer() public {
        Dummy dummy = new DummyDeployer().deploy("salt");
        console.log("Message:", dummy.MESSAGE());
    }
}

contract Dummy {
    string public constant MESSAGE = "Hello";
}

contract DummyDeployer {
    bytes32 immutable internal codeHash = address(new Dummy()).codehash;

    function deploy(bytes32 salt) public returns (Dummy) {
        return Dummy(CONTRACT_DEPLOYER.create2(salt, codeHash, ""));
    }
}
```
The output:
```
[FAIL: revert: This method require system call flag] testDummyDeployer() (gas: 57250803)
Traces:
  [57250803] DummyDeployerTest::testDummyDeployer()
    ├─ [57008011] → new DummyDeployer@0xF9E9ba9Ed9B96AB918c74B21dD0f1D5f2ac38a30
    │   ├─ [91] → new Dummy@0xf232f12E115391c535FD519B00efADf042fc8Be5
    │   │   └─ ← [Return] 544 bytes of code
    │   └─ ← [Return] 2592 bytes of code
    ├─ [209944] DummyDeployer::deploy(0x73616c7400000000000000000000000000000000000000000000000000000000)
    │   └─ ← [Revert] revert: This method require system call flag
    └─ ← [Revert] revert: This method require system call flag
```

**Reply by dutterbutter** [reply]
Let me try to reproduce

**Reply by dutterbutter** [reply]
Okay I see the issue. 

You cannot call ContractDeployer directly as you are but rather with `SystemContractsCaller` like so: 

```
    function deploy(
        bytes32 salt,
        bytes32 bytecodeHash,
        bytes calldata inputData
    ) external payable returns (address create2Address) {
        (bool success, bytes memory returnData) = SystemContractsCaller
            .systemCallWithReturndata(
                uint32(gasleft()),
                address(DEPLOYER_SYSTEM_CONTRACT),
                uint128(0),
                abi.encodeCall(
                    DEPLOYER_SYSTEM_CONTRACT.create2,
                    (salt, codeHash, inputData)
                )
            );

        if (!success) {
            revert Create2FailedDeployment();
        }

        create2Address = abi.decode(returnData, (address));
        console.log("Contract deployed at:", create2Address);
    }
```

That being said, you can use Create2Factory.sol instead to deploy contracts to deterministic addresses for a more convenient approach like so:

`new MyContract{salt: salt}()` 

Here is a min-repo with all available create2 usages:

https://github.com/dutterbutter/min-zksync-create2/tree/main 

To run and view different approaches [here](https://github.com/dutterbutter/min-zksync-create2/blob/main/test/Create2ZKTest.t.sol).

1. `git clone git@github.com:dutterbutter/min-zksync-create2.git` 
2. `cd min-zksync-create2` 
3. `forge install`
4. Fix lib issue before compiling: https://github.com/dutterbutter/min-zksync-create2/tree/main?tab=readme-ov-file#4-fix-the-compilation-error (note:: this will be resolved soon) 
5. Run tests `forge test --zksync --skip Create2Test.t.sol Counter.t.sol`

Let me know if that resolves your issue.

**Reply by CodeSandwich** [reply]
Thank you for the reply, I didn't expect to get THAT verbose example, that's really helpful! There's a lot to unpack here.

 I agree that using `{salt: ...}` is arguably the easiest thing to use, Solidity just supports it, but in many cases it's infeasible. The problem is that the bytecode hash isn't very stable, so it's tricky to make the addresses predictable. At least in my use case the predictability is the main feature, it's the only reason why I'm using CREATE2, and on top of that my contracts are upgradeable, so the derivation must be consistent across the implementation contract deployments. This is a huge problem in plain Solidity because changing a single line of code of the contract or its dependencies or any change in the toolchain like upgrading solc is by design guaranteed to change the hash. In zkSync the problem is even worse than that because the ZK compiler version is not even configurable and comes pre-packaged with the toolkit, in my case foundry. This means that the bytecode hash is as rolling as the zkSync fork of foundry unless we all agree to use the same frozen version forever, which is not feasible in the current state of tooling flux. YMMV, but IMO Solidity's CREATE2 in zkSync is currently borderline useless, you'd need a non-upgradeable contract and ideally no need for external knowledge about the derivation except that it won't collide. (It would be really cool if the hand-crafted contracts like clone proxy or CREATE3 worked in zkSync, that's where CREATE2 really shines!)

I don't think that the example of obtaining  the bytecode hash you provided in `testCreate2SyntaxDeployment` is feasible for production. Sure, it's a very clever piece of code with analyzing the build artifacts, but because it relies on `vm`, it can't be used in production, it's a cool toy to showcase in the test environment.

It's unclear to me when exactly the user will need to provide the bytecode to be deployed via CREATE2 attached to the transaction? Is it every time its deployment is triggered or only on the first deployment and on the second one the wallet will somehow know that the bytecode is known to the network and it doesn't need to be attached anymore? How will the user obtain the bytecode?

`testCreate2FactoryDeployment` is a very nice example, I'd need to rework it a bit, but it shows the right tooling. It uses the `EfficientCall` library from `era-contracts` which is a bit problematic. I would love to use `era-contracts` as my dependency, but all the contracts inside are weirdly pinned to different versions of Solidity. Many helper libraries require exactly 0.8.20, some require 0.8.24 and only some files are more liberal. Do you think that you could relax the version requirement and e.g. switch to `^0.8.20` everywhere? Pinning solc to the one right version from a few years ago just to use some helper libraries would clearly be a mistake.

`testSystemContractDeployment` seems to be doing effectively the same thing as `testCreate2FactoryDeployment` minus the generic factory, but it uses very different utilities to trigger the deployment itself, why is that? Are these routines good for different use cases? Is `EfficientCall.rawCall(..., REAL_DEPLOYER_SYSTEM_CONTRACT, ...)` different from `SystemContractsCaller.systemCallWithReturndata(...)`?

Why are you using `ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT.getRawCodeHash(address(counter))` instead of `address(counter).codehash`, does it provide any advantage?

**Reply by CodeSandwich** [reply]
I've also noticed something really bizarre in assembly inlines. It seems that the `call` Yul instruction has a different meaning in zkSync and in EVM. In EVM the first argument is the gas limit, but in zkSync it's the target address to be called, at least when making system calls. Does it mean that all assembly inlines using `call` written for EVM will fail weirdly on zkSync and need to be rewritten? Can I accidentally call a system contract if I'm unlucky when choosing the gas limit? The weirdest part is that zkSync docs don't mention such difference, the first argument is said to be the gas limit.

**Reply by dutterbutter** [reply]
Happy to provide a resource that is helpful! 

> ZK compiler version is not even configurable

You can configure the compiler via `foundry.toml` as described [here](https://github.com/matter-labs/foundry-zksync/tree/main/crates/config#zksync-settings), albeit we are going to release a zksolc policy as described in this [issue](https://github.com/matter-labs/foundry-zksync/issues/759).

> would be really cool if the hand-crafted contracts like clone proxy or CREATE3 worked in zkSync

If I am not mistaken this may be supported when we release bytecode evm equivalency as part of our 2025 roadmap. If you haven't seen that I recommend checking it out [here](https://x.com/zksync/status/1867296802665386334). 

> I don't think that the example of obtaining the bytecode hash you provided in testCreate2SyntaxDeployment is feasible

Definitely agreed, I merely used it to provide an additional example of retrieving bytecodehash. 

> It's unclear to me when exactly the user will need to provide the bytecode to be deployed via CREATE2 attached to the transaction?

Not exactly sure what you mean here. Can you elaborate? 

> I would love to use era-contracts as my dependency, but all the contracts inside are weirdly pinned to different versions of Solidity

This is something I've heard more then once, I will try to escalate this to the protocol team so they may make adjustments for an improved developer experience 🙏 

>Is EfficientCall.rawCall(..., REAL_DEPLOYER_SYSTEM_CONTRACT, ...) different from SystemContractsCaller.systemCallWithReturndata(...)?

Only difference is `EfficientCall` doesnt re-copy the data into memory it passes a fat pointer to the underlying function

> Why are you using ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT.getRawCodeHash(address(counter)) instead of address(counter).codehash, does it provide any advantage?

tbh I havent really used `address(counter).codehash`, I have always defaulted to querying storage rather then making use of `EXTCODEHASH`. ` address(counter).codehash` is certainly nicer, and both still depend on the contract being known (e.g. deployed previously).

**Reply by CodeSandwich** [reply]
> You can configure the compiler via foundry.toml as described [here](https://github.com/matter-labs/foundry-zksync/tree/main/crates/config#zksync-settings), albeit we are going to release a zksolc policy as described in this https://github.com/matter-labs/foundry-zksync/issues/759.

Awesome, I wasn't aware of these options! They improves the builds reproducibility a lot :+1: 

> release bytecode evm equivalency as part of our 2025 roadmap

Again, that's great, I can't wait for it! I don't expect any commitment, but is it more of a January thing or a Q4-if-stars-align thing?

> Not exactly sure what you mean here. Can you elaborate?

Sure. AFAIK when you deploy a contract for the first time to zkSync, you need to attach its bytecode to the transaction that causes the deployment. This only needs to be done once, after that the deployment can be done again by just using the bytecode hash, the zkSync network is expected to know what the contract is supposed to do. If we have a contract like in the example, where a function deploys another contract, at some point somebody must provide the bytecode of the another deployed contract. My question is who and when? It's entirely possible that the another contract is actually deployed for the first time after a long time or even never. Do I need to provide its bytecode when deploying the "parent" contract that will be deploying it? Or does the user who triggers the first deployment need to do that? If so, how are they expected to obtain the bytecode? What's the default behavior and how to do it "right"?

>  I will try to escalate this to the protocol team so they may make adjustments for an improved developer experience 🙏

Yes please, that would be great! It doesn't seem like an awful lot of work, Solidity is very careful about backward compatibility, but it will enable a ton of tooling for many devs.

**Reply by dutterbutter** [reply, marked as answer]
Hopefully this [min-example](https://github.com/dutterbutter/min-example-factory) provides additional clarity when working with factory contracts. 

In that min-example:

1. **Parent Deployment Without Child Bytecode:**  
   In the example, the `ChildFactory` contract (the parent) is deployed first, but we do not attach the child's bytecode to the factory at this stage. The factory constructor only stores the known bytecode hash, not the full bytecode itself.

2. **First Child Deployment (Providing Bytecode):**  
   When we run the `DeployChild` script, you see that only at the moment we call `factory.deployChild(...)` for the first time we use `vmExt.zkUseFactoryDep("Child")`. This command includes the child's bytecode in the transaction. 

3. **Afterward, Only the Hash Is Needed:**  
 When we call `factory.deployChild(...)` again (using the same child code), we no longer need to provide the `vmExt.zkUseFactoryDep("Child")` line. The code hash is known, so the second and subsequent deployments can rely on the previously registered bytecode.

We also have a similar HH [example](https://code.zksync.io/tutorials/native-aa-multisig) and foundry [here](https://foundry-book.zksync.io/zksync-specifics/examples/smart-account) for native multisigs.  

Does that help?

**Reply by CodeSandwich** [reply]
> Does that help?

Yes, thank you! In practice that means that in my use case (a factory deploying copies of a single contract) it makes the most sense to make a dummy deployment of the child contract during the deployment of the factory. It will give me the correct child bytecode hash with no complex steps and will solve the potential UX problems around of the first deployment of the child because the network will be guaranteed to know the bytecode even when the first user triggers the deployment of the first child.

**Reply by CodeSandwich** [reply]
Hi @dutterbutter, regarding
> 4. Fix lib issue before compiling: https://github.com/dutterbutter/min-zksync-create2/tree/main?tab=readme-ov-file#4-fix-the-compilation-error (note:: this will be resolved soon)

It hasn't been fixed yet and `era-contracts` still needs manual modifications, but there's an open PR fixing that: https://github.com/matter-labs/era-contracts/pull/1152 Maybe you know who to poke in order for it to get merged? Both you and I seem to notice the severity of this issue.

**Reply by CodeSandwich** [reply]
Hi @dutterbutter,
Who should I reach out to regarding https://github.com/matter-labs/era-contracts/pull/1152? The repository's main branch saw no new commits for 3 months now, so it doesn't seem actively developed.

**Comment by dutterbutter** [comment]
>I've also noticed something really bizarre in assembly inlines. It seems that the call Yul instruction has a different meaning in zkSync and in EVM. In EVM the first argument is the gas limit, but in zkSync it's the target address to be called, at least when making system calls. Does it mean that all assembly inlines using call written for EVM will fail weirdly on zkSync and need to be rewritten? Can I accidentally call a system contract if I'm unlucky when choosing the gas limit? The weirdest part is that zkSync docs don't mention such difference, the first argument is said to be the gas limit.

**I asked one of our protocol engineers 👇 :** 

By default zksolc compiler has the same behavior as EVM. However to enable zksync-specific features we use so called "eravm extensions". You can read more about them [there](https://github.com/code-423n4/2023-10-zksync/blob/main/docs/VM%20Section/How%20compiler%20works/instructions/extensions/overview.md). These are special features of our compiler, which roughly look like that: if you use a constant address inside the yul call (and the constant is one from the list [here](https://github.com/code-423n4/2023-10-zksync/blob/main/docs/VM%20Section/How%20compiler%20works/instructions/extensions/call.md) , then yes a special opcode will be invoked). Note, that not all opcodes are available for users for the sake of security, but some are. E.g. this is how EfficientCall library works. For security reasons these extensions are disabled by default, to enable them, you need to provide a special flag to our compiler. An example for [foundry zksync](https://github.com/matter-labs/era-contracts/blob/658713f3a25847662cc997408bd4ccc461f6380a/system-contracts/foundry.toml#L14).

To sum it up:
- By default the behavior is 100% identical to EVM.
- If you turn the flag on (this may be necessary when you are developing a custom account for instance), then this special behavior would only be activated, only if  your call uses a constant special address. I.e. just writing call(gas(), to, ... is not enough, the to must be derived from a constant from the list above known at compile time, which never happens in practice since it would be very weird to use the special addresses from above as a constant (these are empty addresses on Ethereum)

**Reply by CodeSandwich** [reply]
Thanks for taking your time to reach out to the protocol devs :+1: Now it makes sense.
